pkg/image/api/helper.go

```diff
@@ -130,89 +129,54 @@ func IsRegistryDockerHub(registry string) bool {
 	var ref DockerImageReference
-	// TODO replace with docker version once docker/docker PR11109 is merged upstream
-	stream, tag, id := parseRepositoryTag(spec)
 
-	repoParts := strings.Split(stream, "/")
-	switch len(repoParts) {
-	case 2:
-		if isRegistryName(repoParts[0]) {
-			// registry/name
-			ref.Registry = repoParts[0]
-			if IsRegistryDockerHub(ref.Registry) {
-				ref.Namespace = DockerDefaultNamespace
-			}
-			if len(repoParts[1]) == 0 {
-				return ref, fmt.Errorf("the docker pull spec %q must be two or three segments separated by slashes", spec)
-			}
-			ref.Name = repoParts[1]
-			ref.Tag = tag
-			ref.ID = id
-			break
-		}
-		// namespace/name
-		ref.Namespace = repoParts[0]
-		if len(repoParts[1]) == 0 {
-			return ref, fmt.Errorf("the docker pull spec %q must be two or three segments separated by slashes", spec)
-		}
-		ref.Name = repoParts[1]
-		ref.Tag = tag
-		ref.ID = id
-		break
-	case 3:
-		// registry/namespace/name
-		ref.Registry = repoParts[0]
-		ref.Namespace = repoParts[1]
-		if len(repoParts[2]) == 0 {
-			return ref, fmt.Errorf("the docker pull spec %q must be two or three segments separated by slashes", spec)
-		}
-		ref.Name = repoParts[2]
-		ref.Tag = tag
-		ref.ID = id
-		break
-	case 1:
-		// name
-		if len(repoParts[0]) == 0 {
-			return ref, fmt.Errorf("the docker pull spec %q must be two or three segments separated by slashes", spec)
-		}
-		ref.Name = repoParts[0]
-		ref.Tag = tag
-		ref.ID = id
-		break
-	default:
-		// Handle multiple segments in form: registry/namespace/namesegment1/namesegment2/.../name
-		// but not this form: http://registry/namespace/name. We don't support using the schema in
-		// the DockerImageReference.
-		if len(repoParts) > 3 && !strings.HasSuffix(repoParts[0], ":") {
-			if len(repoParts[0]) == 0 {
-				return ref, fmt.Errorf("the docker pull spec %q cannot have empty segments", spec)
-			}
-			ref.Registry = repoParts[0]
+	dockerRef, err := reference.ParseNamed(spec)
+	if err != nil {
+		return ref, err
+	}
 
-			if len(repoParts[1]) == 0 {
-				return ref, fmt.Errorf("the docker pull spec %q cannot have empty segments", spec)
-			}
-			ref.Namespace = repoParts[1]
+	if named, ok := dockerRef.(reference.Named); ok {
+		ref.Registry, ref.Name = splitHostname(named.Name())
+	}
 
-			for i, part := range repoParts[2:] {
-				if len(part) == 0 {
-					return ref, fmt.Errorf("the docker pull spec %q cannot have empty segments", spec)
-				}
-				if i > 0 {
-					ref.Name += "/"
-				}
-				ref.Name += part
-			}
+	if named, ok := dockerRef.(reference.NamedTagged); ok {
+		ref.Tag = named.Tag()
+	}
 
-			ref.Tag = tag
-			ref.ID = id
-			break
+	if named, ok := dockerRef.(reference.Canonical); ok {
+		ref.ID = named.Digest().String()
+	}
+
+	// It's not enough just to use the reference.Parse(). We have the fill
+	// ref.Namespace from ref.Name
+
+	repoParts := strings.Split(ref.Name, "/")
+
+	if len(repoParts) >= 2 {
+		if isRegistryName(repoParts[0]) {
```

Comment:
Is this even possible to satisfy? It is with `localhost` in name (e.g. `registry.com/app/localhost/image`). But it looks to me like a valid reference.